common/hsm_encryption.c

```diff
@@ -79,6 +79,23 @@ bool decrypt_hsm_secret(const struct secret *encryption_key,
 	return true;
 }
 
+bool hsm_secret_check_if_encrypted(const char *path, struct stat *hsm_stat,
+				   char **reason, int *exit_code)
+{
+	// We accept also NULL value, and for us means
+	// that we don't care about the stat data
+	if (sizeof(hsm_stat) != sizeof(struct stat *))
```

Comment:
Yes, this doesn't work (it's always true, whatever the value of hsm_stat is).

Reply:
OK, a few things:

As Linus Torvalds says, "always return the error!".  And in this case, there's only one way it can fail, so the caller knows what the error is.  So:

```
/* Returns NULL if check is OK, an error msg if something went wrong */
char *hsm_secretfile_is_encrypted(const char *path, struct stat *hsm_stat, bool *is_encrypted);
```

But two more refinements:
1. No caller uses the hsm_stat argument, remove it.
2. The string here is not very useful, it needs to incorporate strerror, thus be dynamically allocated.

So:

```
/* Returns NULL if check is OK, an error msg (off ctx) if something went wrong */
char *hsm_secretfile_is_encrypted(const tal_t *ctx, const char *path, bool *is_encrypted)
{
	struct stat st;
	if (stat(path, &st) != 0)
		return tal_fmt(ctx, "Could not stat hsm_secret: %s", strerror(errno));

	*is_encrypted = (st.st_size == ENCRYPTED_HSM_SECRET_LEN);
	return NULL;
}
```

Reply:
Thank you for the code improvement I will add these changes :)

In addition, the only reason that I have used the `hsm_stat` is for a case like that!
```
       bool is_encrypted = hsm_secret_check_if_encrypted(hsm_secret_path, &st,
							  &err_msg, &err_code);
	if (!is_encrypted && st.st_size != 32)
		errx(ERROR_HSM_FILE, "Invalid hsm_secret (neither plaintext "
				     "nor encrypted).");
```

I'm always scared to touch the hsm code and I don't know if we need to include these cases too.

A solution that I have in mind is implementing some things like that

```
/* Returns NULL if check is OK, an error msg (off ctx) if something went wrong */
char *hsm_secretfile_is_encrypted_with_stat(const tal_t *ctx, struct stat *hsm_stat, const char *path, bool *is_encrypted)
{
	*is_encrypted = (hsm_stat.st_size == ENCRYPTED_HSM_SECRET_LEN);
	return NULL;
}

/* Returns NULL if check is OK, an error msg (off ctx) if something went wrong */
char *hsm_secretfile_is_encrypted(const tal_t *ctx, const char *path, bool *is_encrypted)
{
	struct stat st;
	if (stat(path, &st) != 0)
		return tal_fmt(ctx, "Could not stat hsm_secret: %s", strerror(errno));

	*is_encrypted = (st.st_size == ENCRYPTED_HSM_SECRET_LEN);
	return hsm_secretfile_is_encrypted_with_stat(ctx, st, path, is_encypted);
}
```

do you think that this case are not useful? I can safely remove it?